src/v/storage/tests/compaction_fuzz_test.cc

```diff
@@ -0,0 +1,237 @@
+#include <seastar/core/sleep.hh>
+#include <seastar/testing/thread_test_case.hh>
+
+#include <absl/container/btree_map.h>
+#include <boost/test/tools/old/interface.hpp>
+#include <boost/test/unit_test.hpp>
+
+#include <exception>
+
+using namespace std::chrono_literals;
+
+namespace {
+ss::logger cmp_testlog("cmp-fuzz");
+} // anonymous namespace
+
+static model::record_batch make_random_batch(
+  model::offset offset,
+  bool empty,
+  model::record_batch_type type,
+  std::vector<std::optional<ss::sstring>> keys,
+  std::vector<std::optional<ss::sstring>> values,
+  int num_records) {
+    BOOST_REQUIRE(keys.size() == values.size());
+    storage::record_batch_builder builder(type, offset);
+    auto to_iobuf = [](std::optional<ss::sstring> x) {
+        std::optional<iobuf> result;
+        if (x.has_value()) {
+            iobuf buf;
+            buf.append(x->data(), x->size());
+            result = std::move(buf);
+        }
+        return result;
+    };
+    if (!empty) {
+        for (int i = 0; i < num_records; i++) {
+            auto key = random_generators::random_choice(keys);
+            auto val = random_generators::random_choice(values);
+            builder.add_raw_kv(to_iobuf(key), to_iobuf(val));
+        }
+    }
+    return std::move(builder).build();
+}
+
+static fragmented_vector<model::record_batch>
+generate_random_record_batches(int num, int cardinality) {
+    fragmented_vector<model::record_batch> result;
+    std::vector<std::optional<ss::sstring>> keys;
+    std::vector<std::optional<ss::sstring>> values;
+    std::vector<model::record_batch_type> types{
+      model::record_batch_type::raft_configuration,
+      model::record_batch_type::raft_data,
+      model::record_batch_type::archival_metadata,
+    };
+    for (int i = 0; i < cardinality; i++) {
+        if (i == 0) {
+            keys.emplace_back(std::nullopt);
+        } else {
+            keys.emplace_back(
+              random_generators::gen_alphanum_string(20, false));
+        }
+        values.emplace_back(random_generators::gen_alphanum_string(20, false));
+    }
+    // Generate actual batches
+    model::offset current{0};
+    for (int i = 0; i < num; i++) {
+        result.emplace_back(make_random_batch(
+          current,
+          false,
+          random_generators::random_choice(types),
+          keys,
+          values,
+          random_generators::get_int(1, 10)));
+        current = model::next_offset(result.back().last_offset());
+    }
+    return result;
+}
+
+/// Offset translator state observed at some point in time
+struct ot_state {
+    std::deque<model::offset> gap_offset;
+    std::deque<int64_t> gap_length;
+};
+
+/// Consumer that builds the map of all non-data
+/// batches!
+struct ot_state_consumer {
+    ss::future<ss::stop_iteration> operator()(model::record_batch rb) {
+        static const auto translation_batches
+          = model::offset_translator_batch_types();
+        if (
```

Comment:
A different condition than `if(rb.header().type != model::record_batch_type::raft_data)`.

Otherwise, compaction placeholder batches may find their way into the `ot_state` with a `gap_length` of 0.